aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -0,0 +1,466 @@
+#pragma once
+
+#include <ATen/core/stack.h>
+#include <c10/util/TypeList.h>
+#include <ATen/core/boxing/kernel_functor.h>
+#include <ATen/core/boxing/kernel_function.h>
+#include <ATen/core/boxing/kernel_lambda.h>
+
+namespace c10 {
+
+namespace detail {
+template<class Type>
+constexpr uint64_t hashType() {
+  return typeid(Type).hash_code()
+      + 1000 * std::is_lvalue_reference<Type>::value
+      + 5000 * std::is_rvalue_reference<Type>::value
+      + 10000 * std::is_const<guts::remove_reference_t<Type>>::value
+      + 15000 * std::is_volatile<guts::remove_reference_t<Type>>::value
+      ;
+}
+template<class TypeList> struct hashTypeList_ final {};
+template<class Head, class... Tail>
+struct hashTypeList_<guts::typelist::typelist<Head, Tail...>> final {
+  static constexpr uint64_t call(uint64_t index) {
+    return 1000000 * index * hashType<Head>() + hashTypeList_<guts::typelist::typelist<Tail...>>::call(index + 1);
```

Comment:
shameless programming competition plug - this is a really bad hash function :) (not that it's too critical here)

you're adding all items with `index*100000*hashType`, so they can easily collide

you probably wanted to make multiplicative hash - in which case you need to multiply the tail. But in that case 1000000 is a really bad constant as it'd overflow 2^64 and very quickly ignore other args. You want a co-prime number with 2^64. Even then, for best results you should take hashes module some large prime number - it lowers changes of saturation from e.g. indices being divisible by 2.

i.e. `(a0 + c*((a1 + c*(a2 ....))%mod))%mod` where `c` is usually larger than a_i and `mod` is a large prime (you can skip large prime part as you don't have int128 multiplication).

you can also look at folly::hash_combine that does bit rotation magic, but it's a bit too much for this case :)

Reply:
Good points. Ignoring this for now because I removed the error checking, but will consider this if I manage to reintroduce it somehow.